modules/runner/src/integrationTest/java/org/apache/ignite/internal/raftsnapshot/ItTableRaftSnapshotsTest.java

```diff
@@ -792,6 +816,116 @@ void testChangeLeaderDuringSnapshotInstallationToLeaderWithEnoughLog() throws Ex
+
+        // Block AppendEntries from being accepted on the follower so that the leader will have to use a snapshot.
+        blockIncomingAppendEntriesAt(followerIndex);
+
+        // Inhibit the MetaStorage on the follower to make snapshots not eligible for installation.
+        WatchListenerInhibitor listenerInhibitor = inhibitMetastorageListenersAt(followerIndex);
+
+        try {
+            // Add some data in a schema that is not yet available on the follower
+            updateTableSchemaAt(leaderIndex);
+            putToTableAt(leaderIndex);
+
+            CountDownLatch installationRejected = installationRejectedLatch();
+            CountDownLatch snapshotInstalled = snapshotInstalledLatch(followerIndex);
+
+            // Force InstallSnapshot to be used.
+            causeLogTruncationOnSolePartitionLeader(leaderIndex);
+
+            assertTrue(installationRejected.await(20, TimeUnit.SECONDS), "Did not see snapshot installation rejection");
+
+            assertThat("Snapshot was installed before unblocking", snapshotInstalled.getCount(), is(not(0L)));
+
+            listenerInhibitor.stopInhibit();
+
+            assertTrue(snapshotInstalled.await(20, TimeUnit.SECONDS), "Did not see a snapshot installed");
+        } finally {
+            listenerInhibitor.stopInhibit();
+        }
+    }
+
+    private void updateTableSchemaAt(int nodeIndex) {
+        cluster.doInSession(nodeIndex, session -> {
+            session.execute(null, "alter table test add column added int");
+        });
+    }
+
+    private void putToTableAt(int nodeIndex) {
+        KeyValueView<Tuple, Tuple> kvView = cluster.node(nodeIndex)
+                .tables()
+                .table("test")
+                .keyValueView();
+        kvView.put(null, Tuple.create().set("key", 1), Tuple.create().set("val", "one"));
+    }
+
+    private void blockIncomingAppendEntriesAt(int nodeIndex) {
+        BlockingAppendEntriesRequestProcessor blockingProcessorOnFollower = installBlockingAppendEntriesProcessor(nodeIndex);
+
+        blockingProcessorOnFollower.startBlocking();
+    }
+
+    private WatchListenerInhibitor inhibitMetastorageListenersAt(int nodeIndex) {
+        IgniteImpl nodeToInhibitMetaStorage = cluster.node(nodeIndex);
+
+        WatchListenerInhibitor listenerInhibitor = WatchListenerInhibitor.metastorageEventsInhibitor(nodeToInhibitMetaStorage);
+        listenerInhibitor.startInhibit();
+
+        return listenerInhibitor;
+    }
+
+    private CountDownLatch installationRejectedLatch() {
+        CountDownLatch installationRejected = new CountDownLatch(1);
+
+        copierLogInspector.addHandler(
+                event -> event.getMessage().getFormattedMessage().startsWith("Metadata not yet available, URI '"),
+                installationRejected::countDown
+        );
+
+        return installationRejected;
+    }
+
+    private BlockingAppendEntriesRequestProcessor installBlockingAppendEntriesProcessor(int nodeIndex) {
```

Comment:
These tests become more and more complicated with time, any ideas of improvements?
Too much reflection for my taste

Reply:
I wrote it as simply as I could. The test scenario is pretty twisted by its virtue. Any suggestions are welcome

Reply:
Can we have some unit tests maybe? We're not even trying to write them for some reason

Reply:
Added a unit test